onnxruntime/core/providers/coreml/builders/impl/activation_op_builder.cc

```diff
@@ -54,6 +110,66 @@
 
 // Operator support related
 
+namespace {
+// assumes that node.OpType() == "PRelu"
+bool IsPReluOpSupported(const Node& node, const OpBuilderInputParams& input_params,
+                        const logging::Logger& logger) {
+  const auto& input_defs = node.InputDefs();
+
+  // X input rank must be at least 3
+  std::vector<int64_t> x_shape;
+  if (!GetShape(*input_defs[0], x_shape, logger)) {
+    return false;
+  }
+  if (x_shape.size() < 3) {
+    LOGS(logger, VERBOSE) << "PRelu 'X' input must have at least 3 dimensions";
+    return false;
+  }
+
+  // slope input must be a constant initializer
+  if (!input_params.graph_viewer.IsConstantInitializer(input_defs[1]->Name(), true)) {
+    LOGS(logger, VERBOSE) << "PRelu 'slope' input must be a constant initializer tensor";
+    return false;
+  }
+
+  // slope must either:
+  // - have shape [C, 1, 1]
+  // - have 1 element
+  {
+    std::vector<int64_t> slope_shape;
+    if (!GetShape(*input_defs[1], slope_shape, logger)) {
+      return false;
+    }
+    const bool has_per_channel_slopes =
+        (slope_shape.size() == 3 && std::all_of(slope_shape.begin() + 1, slope_shape.end(),
+                                                [](int64_t dim) { return dim == 1; }));
+    const bool has_single_slope = Product(slope_shape) == 1;
+    if (!has_per_channel_slopes && !has_single_slope) {
+      LOGS(logger, VERBOSE) << "PRelu 'slope' input must either have shape [C, 1, 1] or have a single value";
+      return false;
+    }
+
+    if (has_single_slope && x_shape[x_shape.size() - 3] == 1) {
+      // TODO: CoreML crashes with single element slope, hence this special case. Remove when fixed.
+      // https://github.com/apple/coremltools/issues/1488
+      LOGS(logger, VERBOSE) << "PRelu single 'slope' value in CoreML weight is not supported";
+      return false;
+    }
+  }
```

Comment:
Not quite understanding this.

If AddPReluWeight can do the broadcast of a single value what is this check for? Also not quite understanding the check on x_shape.size() - 3 given rank could be 3 or more couldn't it?

Reply:
> Does it crash if there's a single slope element and C==1?

Yes, that also crashes.

> Also not quite understanding the check on x_shape.size() - 3 given rank could be 3 or more couldn't it? Maybe the coreml spec is wrong as that's typically NCHW so the C channel being at index -3 if the input was rank 5 seems off.

That's the check for dimension "C" from the CoreML spec. I don't quite understand why they specify C at index -3 either.

Reply:
Added check for X's rank being 3 or 4.